File: envoy/envelope-session-demo/backend.py
#!/usr/bin/env python3

from flask import Flask, request, jsonify, make_response
import uuid
import json
import base64

app = Flask(__name__)

# Simple in-memory session store
sessions = {}

def parse_envelope_session(session_header):
    """Parse envelope session header format: base64(json({session_id, ...}))"""
    try:
        decoded = base64.b64decode(session_header).decode('utf-8')
        envelope = json.loads(decoded)
        return envelope
    except Exception as e:
        print(f"Failed to parse envelope session: {e}")
        return None

def create_envelope_session(session_id, additional_data=None):
    """Create envelope session header format: base64(json({session_id, ...}))"""
    envelope = {"session_id": session_id}
    if additional_data:
        envelope.update(additional_data)
    
    json_str = json.dumps(envelope)
    encoded = base64.b64encode(json_str.encode('utf-8')).decode('utf-8')
    return encoded

def is_envelope_format(session_header):
    """Check if session header is in envelope format (base64 encoded JSON)"""
    try:
        decoded = base64.b64decode(session_header).decode('utf-8')
        json.loads(decoded)
        return True
    except:
        return False

@app.route('/')
def home():
    session_header = request.headers.get('x-session-id')
    
    if session_header:
        session_id = None
        
        # Check if it's envelope format or plain header format
        if is_envelope_format(session_header):
            # Envelope format
            envelope = parse_envelope_session(session_header)
            if envelope and 'session_id' in envelope:
                session_id = envelope['session_id']
                
                if session_id in sessions:
                    sessions[session_id]['request_count'] += 1
                    response_data = {
                        "message": f"Hello! Continuing envelope session {session_id}",
                        "session_id": session_id,
                        "format": "envelope",
                        "envelope": envelope,
                        "request_count": sessions[session_id]['request_count'],
                        "session_data": sessions[session_id]
                    }
                    
                    # Update envelope with new request count
                    new_envelope = create_envelope_session(
                        session_id, 
                        {"request_count": sessions[session_id]['request_count']}
                    )
                    
                    response = make_response(jsonify(response_data))
                    response.headers['x-session-id'] = new_envelope
                    return response
        else:
            # Plain header format
            session_id = session_header
            if session_id in sessions:
                sessions[session_id]['request_count'] += 1
                response_data = {
                    "message": f"Hello! Continuing header session {session_id}",
                    "session_id": session_id,
                    "format": "header",
                    "request_count": sessions[session_id]['request_count'],
                    "session_data": sessions[session_id]
                }
                
                response = make_response(jsonify(response_data))
                response.headers['x-session-id'] = session_id
                return response
    
    # Create new session
    session_id = str(uuid.uuid4())
    sessions[session_id] = {
        "created_at": "now",
        "request_count": 1,
        "user_data": {}
    }
    
    # Check if we should use envelope format (for testing)
    use_envelope = request.args.get('envelope', 'false').lower() == 'true'
    
    if use_envelope:
        envelope_session = create_envelope_session(session_id, {"request_count": 1})
        response_data = {
            "message": f"Hello! New envelope session created: {session_id}",
            "session_id": session_id,
            "format": "envelope",
            "request_count": 1
        }
        response = make_response(jsonify(response_data))
        response.headers['x-session-id'] = envelope_session
    else:
        response_data = {
            "message": f"Hello! New header session created: {session_id}",
            "session_id": session_id,
            "format": "header",
            "request_count": 1
        }
        response = make_response(jsonify(response_data))
        response.headers['x-session-id'] = session_id
    
    return response

@app.route('/status')
def status():
    session_header = request.headers.get('x-session-id')
    
    response_data = {
        "endpoint": "/status",
        "session_header_present": bool(session_header),
        "session_header": session_header,
        "total_sessions": len(sessions)
    }
    
    if session_header:
        if is_envelope_format(session_header):
            envelope = parse_envelope_session(session_header)
            response_data["format"] = "envelope"
            response_data["envelope"] = envelope
            if envelope and 'session_id' in envelope:
                response_data["session_id"] = envelope['session_id']
                response_data["session_valid"] = envelope['session_id'] in sessions
        else:
            response_data["format"] = "header"
            response_data["session_id"] = session_header
            response_data["session_valid"] = session_header in sessions
    
    return jsonify(response_data)

@app.route('/data', methods=['GET', 'POST'])
def data():
    session_header = request.headers.get('x-session-id')
    session_id = None
    
    # Extract session ID regardless of format
    if session_header:
        if is_envelope_format(session_header):
            envelope = parse_envelope_session(session_header)
            if envelope and 'session_id' in envelope:
                session_id = envelope['session_id']
        else:
            session_id = session_header
    
    if request.method == 'POST':
        data = request.get_json() or {}
        
        # Store data in session if session exists
        if session_id and session_id in sessions:
            sessions[session_id]['user_data'].update(data)
            sessions[session_id]['request_count'] += 1
            
        response_data = {
            "method": "POST",
            "received_data": data,
            "session_id": session_id,
            "session_stored": session_id in sessions if session_id else False,
            "message": f"Data received for session {session_id}" if session_id else "Data received (no session)"
        }
    else:
        response_data = {
            "method": "GET",
            "session_id": session_id,
            "sample_data": {"key": "value", "number": 42},
            "session_data": sessions.get(session_id, {}) if session_id else {},
            "message": f"Data for session {session_id}" if session_id else "Data (no session)"
        }
        
        # Increment request count for valid sessions
        if session_id and session_id in sessions:
            sessions[session_id]['request_count'] += 1
    
    response = make_response(jsonify(response_data))
    
    # Pass through session header
    if session_header:
        response.headers['x-session-id'] = session_header
    
    return response

@app.route('/sessions')
def list_sessions():
    """List all active sessions"""
    return jsonify({
        "active_sessions": len(sessions),
        "sessions": {k: v for k, v in sessions.items()}
    })

if __name__ == '__main__':
    print("Starting Flask backend server...")
    print("Supports both header-based and envelope session formats:")
    print("  Header format: x-session-id: <session-uuid>")
    print("  Envelope format: x-session-id: base64(json({session_id: <uuid>}))")
    print("")
    print("Test commands:")
    print("  # Header-based sessions (default):")
    print("  curl http://localhost:8080/")
    print("  curl http://localhost:8080/status")
    print("  curl http://localhost:8080/sessions")
    print("")
    print("  # Envelope sessions (add ?envelope=true):")
    print("  curl 'http://localhost:8080/?envelope=true'")
    print("  curl -X POST -H 'Content-Type: application/json' -d '{\"test\": \"data\"}' http://localhost:8080/data")
    print()

@app.route('/health', methods=['GET'])
def health():
    """Health check endpoint for Docker"""
    return jsonify({"status": "healthy", "service": "backend"}), 200

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=8000, debug=True)

File: envoy/stateful-session/envelope/backend.go
package main

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"sync"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)

type Session struct {
	CreatedAt    string                 `json:"created_at"`
	RequestCount int                    `json:"request_count"`
	UserData     map[string]interface{} `json:"user_data"`
}

type Envelope struct {
	SessionID    string `json:"session_id"`
	RequestCount int    `json:"request_count,omitempty"`
}

var (
	sessions   = make(map[string]*Session)
	sessionsMu sync.Mutex
)

func parseEnvelopeSession(sessionHeader string) (*Envelope, error) {
	decoded, err := base64.StdEncoding.DecodeString(sessionHeader)
	if err != nil {
		return nil, err
	}
	var envelope Envelope
	if err := json.Unmarshal(decoded, &envelope); err != nil {
		return nil, err
	}
	return &envelope, nil
}

func createEnvelopeSession(sessionID string, additional map[string]interface{}) (string, error) {
	envelope := map[string]interface{}{
		"session_id": sessionID,
	}
	for k, v := range additional {
		envelope[k] = v
	}
	jsonBytes, err := json.Marshal(envelope)
	if err != nil {
		return "", err
	}
	return base64.StdEncoding.EncodeToString(jsonBytes), nil
}

func isEnvelopeFormat(sessionHeader string) bool {
	decoded, err := base64.StdEncoding.DecodeString(sessionHeader)
	if err != nil {
		return false
	}
	var js map[string]interface{}
	return json.Unmarshal(decoded, &js) == nil
}

func main() {
	r := gin.Default()

	r.GET("/", func(c *gin.Context) {
		sessionHeader := c.GetHeader("x-session-id")
		var sessionID string
		var format string
		var envelope *Envelope

		sessionsMu.Lock()
		defer sessionsMu.Unlock()

		if sessionHeader != "" {
			if isEnvelopeFormat(sessionHeader) {
				format = "envelope"
				env, err := parseEnvelopeSession(sessionHeader)
				if err == nil && env.SessionID != "" {
					sessionID = env.SessionID
					envelope = env
					if sess, ok := sessions[sessionID]; ok {
						sess.RequestCount++
						additional := map[string]interface{}{"request_count": sess.RequestCount}
						newEnvelope, _ := createEnvelopeSession(sessionID, additional)
						c.Header("x-session-id", newEnvelope)
						c.JSON(http.StatusOK, gin.H{
							"message":       fmt.Sprintf("Hello! Continuing envelope session %s", sessionID),
							"session_id":    sessionID,
							"format":        format,
							"envelope":      envelope,
							"request_count": sess.RequestCount,
							"session_data":  sess,
						})
						return
					}
				}
			} else {
				format = "header"
				sessionID = sessionHeader
				if sess, ok := sessions[sessionID]; ok {
					sess.RequestCount++
					c.Header("x-session-id", sessionID)
					c.JSON(http.StatusOK, gin.H{
						"message":       fmt.Sprintf("Hello! Continuing header session %s", sessionID),
						"session_id":    sessionID,
						"format":        format,
						"request_count": sess.RequestCount,
						"session_data":  sess,
					})
					return
				}
			}
		}

		// Create new session
		sessionID = uuid.New().String()
		sessions[sessionID] = &Session{
			CreatedAt:    "now",
			RequestCount: 1,
			UserData:     map[string]interface{}{},
		}
		useEnvelope := strings.ToLower(c.Query("envelope")) == "true"
		if useEnvelope {
			envelopeSession, _ := createEnvelopeSession(sessionID, map[string]interface{}{"request_count": 1})
			c.Header("x-session-id", envelopeSession)
			c.JSON(http.StatusOK, gin.H{
				"message":       fmt.Sprintf("Hello! New envelope session created: %s", sessionID),
				"session_id":    sessionID,
				"format":        "envelope",
				"request_count": 1,
			})
		} else {
			c.Header("x-session-id", sessionID)
			c.JSON(http.StatusOK, gin.H{
				"message":       fmt.Sprintf("Hello! New header session created: %s", sessionID),
				"session_id":    sessionID,
				"format":        "header",
				"request_count": 1,
			})
		}
	})

	r.GET("/status", func(c *gin.Context) {
		sessionHeader := c.GetHeader("x-session-id")
		resp := gin.H{
			"endpoint":               "/status",
			"session_header_present": sessionHeader != "",
			"session_header":         sessionHeader,
			"total_sessions":         len(sessions),
		}
		if sessionHeader != "" {
			if isEnvelopeFormat(sessionHeader) {
				env, _ := parseEnvelopeSession(sessionHeader)
				resp["format"] = "envelope"
				resp["envelope"] = env
				if env != nil && env.SessionID != "" {
					resp["session_id"] = env.SessionID
					resp["session_valid"] = sessions[env.SessionID] != nil
				}
			} else {
				resp["format"] = "header"
				resp["session_id"] = sessionHeader
				resp["session_valid"] = sessions[sessionHeader] != nil
			}
		}
		c.JSON(http.StatusOK, resp)
	})

	r.GET("/data", func(c *gin.Context) {
		sessionHeader := c.GetHeader("x-session-id")
		var sessionID string
		if sessionHeader != "" {
			if isEnvelopeFormat(sessionHeader) {
				env, _ := parseEnvelopeSession(sessionHeader)
				if env != nil {
					sessionID = env.SessionID
				}
			} else {
				sessionID = sessionHeader
			}
		}
		resp := gin.H{
			"method":      "GET",
			"session_id":  sessionID,
			"sample_data": gin.H{"key": "value", "number": 42},
			"session_data": func() interface{} {
				if sessionID != "" {
					return sessions[sessionID]
				}
				return gin.H{}
			}(),
			"message": func() string {
				if sessionID != "" {
					return fmt.Sprintf("Data for session %s", sessionID)
				}
				return "Data (no session)"
			}(),
		}
		if sessionID != "" && sessions[sessionID] != nil {
			sessions[sessionID].RequestCount++
		}
		c.Header("x-session-id", sessionHeader)
		c.JSON(http.StatusOK, resp)
	})

	r.POST("/data", func(c *gin.Context) {
		sessionHeader := c.GetHeader("x-session-id")
		var sessionID string
		if sessionHeader != "" {
			if isEnvelopeFormat(sessionHeader) {
				env, _ := parseEnvelopeSession(sessionHeader)
				if env != nil {
					sessionID = env.SessionID
				}
			} else {
				sessionID = sessionHeader
			}
		}
		var data map[string]interface{}
		if err := c.BindJSON(&data); err != nil {
			data = map[string]interface{}{}
		}
		stored := false
		if sessionID != "" && sessions[sessionID] != nil {
			for k, v := range data {
				sessions[sessionID].UserData[k] = v
			}
			sessions[sessionID].RequestCount++
			stored = true
		}
		resp := gin.H{
			"method":         "POST",
			"received_data":  data,
			"session_id":     sessionID,
			"session_stored": stored,
			"message": func() string {
				if sessionID != "" {
					return fmt.Sprintf("Data received for session %s", sessionID)
				}
				return "Data received (no session)"
			}(),
		}
		c.Header("x-session-id", sessionHeader)
		c.JSON(http.StatusOK, resp)
	})

	r.GET("/sessions", func(c *gin.Context) {
		sessionsMu.Lock()
		defer sessionsMu.Unlock()
		c.JSON(http.StatusOK, gin.H{
			"active_sessions": len(sessions),
			"sessions":        sessions,
		})
	})

	r.GET("/health", func(c *gin.Context) {
		c.JSON(http.StatusOK, gin.H{
			"status":  "healthy",
			"service": "backend",
		})
	})

	fmt.Println("Starting Gin backend server...")
	fmt.Println("Supports both header-based and envelope session formats:")
	fmt.Println("  Header format: x-session-id: <session-uuid>")
	fmt.Println("  Envelope format: x-session-id: base64(json({session_id: <uuid>}))")
	fmt.Println("")
	fmt.Println("Test commands:")
	fmt.Println("  # Header-based sessions (default):")
	fmt.Println("  curl http://localhost:8000/")
	fmt.Println("  curl http://localhost:8000/status")
	fmt.Println("  curl http://localhost:8000/sessions")
	fmt.Println("")
	fmt.Println("  # Envelope sessions (add ?envelope=true):")
	fmt.Println("  curl 'http://localhost:8000/?envelope=true'")
	fmt.Println("  curl -X POST -H 'Content-Type: application/json' -d '{\"test\": \"data\"}' http://localhost:8000/data")
	fmt.Println("")

	r.Run("0.0.0.0:8000")
}
